### DISCUSSION TITLE
Code stops running while using color_topusage()

### DISCUSSION THREAD

**Discussion by meghanaviyyapu** [pymupdf/PyMuPDF#2575] [Looking for help] [answered]
Have used the below PDF for identifying background color of the bbox provided in code. The code execution is failing. Can you please let me
[bookmarks.pdf](https://github.com/pymupdf/PyMuPDF/files/12237365/bookmarks.pdf)
 know how the output for the above method can be extracted.

**Comment by meghanaviyyapu** [comment]
import fitz
doc=fitz.open(file)
clip=(451.9859619140625,792.6179809570312,554.010009765625,817.3499755859375)
page=doc[0]
pix=page.get_pixmap(clip=clip)
x=pix.color_topusage()

**Reply by JorjMcKie** [reply]
The given rectangle is outside the page!
It is a rotated A4 page (90°). Therefore, y-coordinates larger than 595 as in your rect example are outside.
Unfortunately, the method does not check the validity of a clip and produces a Python interpreter crash.
I will change that in the next version.

**Reply by JorjMcKie** [reply]
I recommend checking `clip in pix.irect`.

**Reply by meghanaviyyapu** [reply]
Thanks should bbox values be converted using pix.irect?

**Reply by JorjMcKie** [reply]
The pixmap knows nothing about the page rotation. So its irect is derived from page.rect.
Your rectangle was probably derived from text extraction of the page, and therefore has **_unrotated_** coordinates.
To avoid confusing handling of (de-) rotations, I recommend the following approach if you fear you have rotated pages:
* make a new version of your PDF that
   1. looks exaxtly the same
   2. has unrotated pages
   3. locate desired text (via search etc.)
   4. get background color or what ever you desire

Here is a snippet that does that, then searches for "Lawrence of Arabia" on page 0 and determines th background color:
```python
import fitz
from pprint import pprint
src=fitz.open("bookmarks.pdf")
doc = fitz.open()
for page in src:
    rot=page.rotation
    w,h=page.rect.br
    page.set_rotation(0)
    npage=doc.new_page(width=w,height=h)
    npage.show_pdf_page(npage.rect,src,page.number,rotate=-rot)
src.close()

page=doc[0]
rl = page.search_for("Lawrence of Arabia")
clip=rl[0]
print(clip)
pix = page.get_pixmap()

cc=pix.color_topusage(clip=clip)
pprint(cc)
page.draw_rect(clip,color=(1,0,0))  # demo only
doc.ez_save("rotate.pdf")  # demo only
```
Note:
The script works the same also, if pages are not rotated.

**Comment by meghanaviyyapu** [comment]
Thanks for explaining and the code.

My objective is to take span["bbox"] values to identify background color of text. Could you let me know if it is possible without making a new version of PDF and using page.search_for()?

**Reply by JorjMcKie** [reply, marked as answer]
It is of no importance, from where you get the bbox whose background color you need.
My code above only shows how to avoid the fairly confusing handling of page rotations. But you can also do this:
```python
import fitz
from pprint import pprint

doc = fitz.open("bookmarks.pdf")
page = doc[0]
pix = page.get_pixmap()
blocks = page.get_text("dict")["blocks"]
for b in blocks:
    for l in b["lines"]:
        for s in l["spans"]:
            if s["text"] == "Lawrence of Arabia":
                # adjust bbox to pixmap coordinates:
                bbox = fitz.Rect(s["bbox"]) * page.rotation_matrix
                pprint(pix.color_topusage(clip=bbox))
                print()
```

**Reply by meghanaviyyapu** [reply]
Ok Thanks. Will try using this

**Comment by meghanaviyyapu** [comment]
Can you let me know how a bounding box can be drawn around rotated text in a PDF. For example, in the above PDF, "_The Majestic_" is rotated.

**Reply by JorjMcKie** [reply]
```python
import fitz
doc=fitz.open("bookmarks.pdf")
page=doc[0]
page.clean_contents()
bb=page.search_for("The Majestic")[0]
bb
Rect(451.9859619140625, 792.6179809570312, 554.010009765625, 817.3499755859375)
page.draw_rect(bb, color=(1, 0, 0))
doc.save("x.pdf")
```
Gives you this:
![grafik](https://github.com/pymupdf/PyMuPDF/assets/8290722/f390b6bb-cf40-4bf9-b2fa-cce9a8c21741)

**Reply by meghanaviyyapu** [reply]
Thanks a lot. Without using page.clean_contents() the bbox was getting drawn at a different location. May I know if it is necessary to use page.clean_contents()?

**Reply by JorjMcKie** [reply]
Sometimes yes.
Background: see the documentation [here](https://pymupdf.readthedocs.io/en/latest/recipes-common-issues-and-their-solutions.html): **"Misplaced Item Insertions on PDF Pages"**.
